openssl/src/cms.rs

```diff
@@ -202,11 +281,161 @@ impl CmsContentInfo {
+    use x509::X509NameRef;
+    use foreign_types::Opaque;
+    use super::*;
+
+    impl CmsContentInfoRef {
+        /// Given that we're dealing with the `EnvelopedData`, returns the recipients of the message.
+        ///
+        /// OpenSSL documentation at [`CMS_get0_RecipientInfos`]
+        ///
+        /// [`CMS_get0_RecipientInfos`]: https://www.openssl.org/docs/manmaster/man3/CMS_RecipientInfo_decrypt.html
+        pub fn get_recipient_infos(&self) -> Result<&StackRef<CmsRecipientInfo>, ErrorStack> {
+            unsafe {
+                let recipient_infos = cvt_p(ffi::CMS_get0_RecipientInfos(self.as_ptr()))?;
+                Ok(StackRef::from_ptr(recipient_infos))
+            }
+        }
+    }
+
+    /// An owned version of [`CmsRecipientInfoRef`]
+    ///
+    /// [`CmsRecipientInfoRef`]:struct.CmsRecipientInfoRef.html
+    pub struct CmsRecipientInfo(*mut ffi::CMS_RecipientInfo);
+
+
+    impl ForeignType for CmsRecipientInfo {
+        type CType = ffi::CMS_RecipientInfo;
+        type Ref = CmsRecipientInfoRef;
+
+        #[inline]
+        unsafe fn from_ptr(ptr: *mut Self::CType) -> Self {
+            CmsRecipientInfo(ptr)
+        }
+
+        #[inline]
+        fn as_ptr(&self) -> *mut Self::CType {
+            self.0
+        }
+    }
+
+    /// Reference to a [`CMS_RecipientInfo`].
+    ///
+    /// [`CMS_RecipientInfo`]: https://www.openssl.org/docs/manmaster/man3/CMS_RecipientInfo_decrypt.html
+    pub struct CmsRecipientInfoRef(Opaque);
+
+    impl ForeignTypeRef for CmsRecipientInfoRef {
+        type CType = ffi::CMS_RecipientInfo;
+    }
+
+    impl Stackable for CmsRecipientInfo {
+        type StackType = ffi::stack_st_CMS_RecipientInfo;
+    }
+
+    /// The bindings to the openssl's type of `CMS_RecipientInfo` which can be extracted with
+    /// openssl's `CMS_RecipientInfo_ktri_get0_singer_id()` function.
+    pub enum RecipientInfo<'cms> {
+        Identifier,
+        Info {
+            issuer: &'cms X509NameRef,
+            serial_number: &'cms Asn1IntegerRef,
+        },
+    }
+
+    impl CmsRecipientInfoRef {
+        /// Implementation of [`CMS_RecipientInfo_ktri_get0_signer_id`].
+        ///
+        /// [`CMS_RecipientInfo_ktri_get0_signer_id`]: https://www.openssl.org/docs/manmaster/man3/CMS_RecipientInfo_decrypt.html
+        pub fn get_recipient_info<'cms>(&self) -> Result<RecipientInfo<'cms>, ErrorStack> {
+            unsafe {
+                let mut key_id = ptr::null_mut();
+                let mut issuer = ptr::null_mut();
+                let mut serial_number = ptr::null_mut();
+                cvt(ffi::CMS_RecipientInfo_ktri_get0_signer_id(
+                    self.as_ptr(),
+                    &mut key_id,
+                    &mut issuer,
+                    &mut serial_number,
+                ))?;
+
+                if !key_id.is_null() {
+                    // TODO: Implement (ffi::ASN1_OCTET_STRING) bindings.
+                    Ok(RecipientInfo::Identifier)
```

Comment:
We should at least define the Asn1OctetString type so the enum can hold it.

Reply:
Well, there is some opaque type defined in `ossl_typ.rs`

```
openssl-sys/src/ossl_typ.rs
13:pub enum ASN1_OCTET_STRING {}
```

Not sure if I should return it as it does not hold a pointer by itself. I may of course create some sort of a temporary duplicate which just holds a pointer, but I don't know if that would be ok though.